### PR TITLE
Implement a custom spatial index

### LINKED ISSUE #15082: Implement a custom spatial index
Currently we rely on flatbush for spatial indexing which is more tailored towards maps and therefore leaves some performance on the table.
Given that the spatial indexing core is rather short (albeit complex) the idea would be have a custom implementation and get rid of the additional dependency.

### PR DESCRIPTION
**Pull request by muendlein**
- [x] issues: fixes #15082
- [x] tests added / passed

Panning performance (1M random items):
Speed up for the search function (1.5-2x), overall speed for panning: 20-33% (40 ms -> 30 ms with Chrome & Win11)

Firefox (before)
firefox_before
Firefox (after)
firefox_after

Chrome (before)
chrome_before

Chrome (after)
chrome_after

### CODE REVIEW

**Comment by bryevdv** [comment]
It would be very helpful to also have perf comparisons for FF and Safari, as well as before/after bundle size change. 

Otherwise, I guess my main concern (modulo existing test failures to figure out) is that presumably flatbush has unit test suite to cover actual index implementation correctness. If we are going to adopt a small but complex component I'd really like to see a unit test suite of some sort. Perhaps this is grunt work Claude or Codex could help with.

**Comment by muendlein** [comment]
@bryevdv FF is already there (in form of screenshots), I will follow up with some more benchmarking including Safari (also including indexing performance). There is also more room for future improvement, e.g. specialized index for point based data, setting indices by range instead of individual bits, etc.
As for the bundle size change, there is basically none (< 1 KB).

The unit test suite from flatbush is unfortunately rather limited, so I will add relevant tests soon. The spatial index is at least already tested indirectly as it is it utilized by many other tests.

File: bokehjs/src/lib/core/util/spatial.ts [hunk 1]
```diff
@@ -1,40 +1,256 @@
...
+  private minY: number
+  private maxX: number
+  private maxY: number
+
+  // buffer layout: [coordinate rects - coordinate indices]
+  private _data_byte_buffer: ArrayBuffer
+  private _bboxes: TypedArray
+  private _indices: Uint16Array | Uint32Array
+
+  constructor(n_items: number, node_size: number = 16, array_type_coordinates: TypedArrayConstructor = Float64Array) {
```
Comment (mattpap) [inline bokehjs/src/lib/core/util/spatial.ts:126]: It's fine to be generic, but is there a realistic scenario that we would need anything else than `Float64Array` in this context?

File: bokehjs/src/lib/core/util/spatial.ts [hunk 2]
```diff
@@ -1,40 +1,256 @@
...
+  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
+  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
+  OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
+  TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF
+  THIS SOFTWARE.
+  */
+
+  // 4 coordinates values per rect -> bitshift by 2 (or multiply by 4) to transform index
+  private readonly factor_bbox: number = 4
+  private readonly shift_factor_bbox: number = 2
```
Comment (mattpap) [inline bokehjs/src/lib/core/util/spatial.ts:106]: This really ought to be `static`, but then you won't be able to use `this` to access these attributes, so I won't complain if this is kept as-is.

File: bokehjs/src/lib/core/util/spatial.ts [hunk 3]
```diff
@@ -96,4 +299,251 @@ export class SpatialIndex {
     })
     return result
   }
+
+  search(
+    minX: number,
+    minY: number,
+    maxX: number,
+    maxY: number,
+    leaf_fn: (index: number, x0: number, y0: number, x1: number, y1: number) => boolean,
```
Comment (mattpap) [inline bokehjs/src/lib/core/util/spatial.ts]: It see this type being repeated, so let's factor it out as `type LeafFn = ...`. As a bonus, you won't need to spread this signature so much.

File: bokehjs/src/lib/core/util/spatial.ts [hunk 4]
```diff
@@ -96,4 +299,251 @@ export class SpatialIndex {
...
+      return
+    }
+
+    if (_coordinate_index_position !== _bboxes.length) {
+      throw new Error("Data is not yet indexed - call finish().")
+    }
+
+    const node_index = _bboxes.length - this.factor_bbox
+
+    this._searchRecursive(minX, minY, maxX, maxY, node_index, leaf_fn)
```
Comment (mattpap) [inline bokehjs/src/lib/core/util/spatial.ts]: I would prefer if we kept a consistent naming convention. We really only allow non-snake case conventions, when we inherit them from the environment or dependencies.

**Comment by mattpap** [comment]
Overall looks good. As Bryan mentioned, a test suite is needed.

File: bokehjs/src/lib/core/util/spatial.ts (hunk 1 above)
Comment (muendlein) [reply bokehjs/src/lib/core/util/spatial.ts:126]: Long term actually yes, namely in case of numpy arrays and which could become relevant when typed arrays are more utilized instead of `number[]`. Additionally one could also rescale and downcast the data to float16 or float32 as losing a bit of precision does not really alter the search results (bbox precision is limited by pixel precision anyway).
Overall, reducing the number of bytes will be helpful not only for searching but especially for indexing itself, as the algorithm is quite cache/memory intense.

File: bokehjs/src/lib/core/util/spatial.ts (hunk 3 above)
Comment (muendlein) [reply bokehjs/src/lib/core/util/spatial.ts]: Agreed.

File: bokehjs/src/lib/core/util/spatial.ts (hunk 4 above)
Comment (muendlein) [reply bokehjs/src/lib/core/util/spatial.ts]: Definitely, this was only by mistake as I took things over from my upstream PR (flatbush uses snake case).

File: bokehjs/src/lib/core/util/spatial.ts [hunk 5]
```diff
@@ -1,85 +1,279 @@
...
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ SOFTWARE.
+
+ LUT-based Hilbert curve: xy → 32-bit Hilbert index on a 2^16 × 2^16 grid.
+
+ The 2D Hilbert curve is a state machine with 4 orientations.
+ At each step we consume 4 bits of x and 4 bits of y (8 bits total = 1 Morton byte),
+ emit 8 bits of Hilbert output, and transition to the next state.
+ 16 input bits ÷ 4 bits/step = 4 steps total (fully unrolled).
+*/
```
Comment (bryevdv) [inline bokehjs/src/lib/core/util/spatial.ts:36]: Is there an academic paper or existing implementation that could be referenced here?

Comment (muendlein) [reply bokehjs/src/lib/core/util/spatial.ts:36]: Not sure about academic papers but the existing implementation that are used as a baseline are mentioned at the beginning of the comment. I have now also included direct repo links to make this explicite.

**Comment by muendlein** [comment]
@bryevdv It seems like that unit tests are already in place: https://github.com/bokeh/bokeh/blob/branch-3.10/bokehjs/test/unit/core/util/spatial.ts
Do you think this should be extended further?

**Comment by muendlein** [comment]
Here are the performance metrics (2M random points).
Note that Safari is not included as it does not seem to have a proper profiler (no flame charts, incomplete broken call trees).

| Metric | Browser | Before | After |
| :--- | :--- | :--- | :--- |
| **Indexing Performance** | Firefox | 600 ms | 485 ms |
| **Indexing Performance** | Chrome | 430 ms | 405 ms |
| **Search Performance** | Firefox | 65 ms | 38 ms |
| **Search Performance** | Chrome | 40 ms | 17 ms |

**Comment by bryevdv** [comment]
> Do you think this should be extended further?

@muendlein those are a good start I had forgotten about them. But this PR adds a public `search` method. I think that needs tests at a minimum but ideally also some for `compute_hilbert` as well, since it is exported. 

Edit: alternatively, those could be made private, I suppose.

**Review by bryevdv — APPROVED** [review]
LGTM modulo one remaining comment re: tests

**Comment by muendlein** [comment]
> > Do you think this should be extended further?
> 
> @muendlein those are a good start I had forgotten about them. But this PR adds a public `search` method. I think that needs tests at a minimum but ideally also some for `compute_hilbert` as well, since it is exported.
> 
> Edit: alternatively, those could be made private, I suppose.

@bryevdv 
Probably better to make them private as I don't expect them to be used somewhere else.
